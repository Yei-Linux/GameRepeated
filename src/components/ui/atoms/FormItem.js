import Container from '../layouts/Container'
import Input from './Input'
import Label from './Label'

const FormItem = ({
  testId,
  htmlFor,
  ariaLabel,
  type = 'input',
  className,
  classNameInput,
  label = '',
  name,
  onChange,
  value,
}) => {
  const elementTypes = {
    input: Input,
  }

  const Element = elementTypes[type]

  const handleChange = (e) => {
    const valueEvent = e.target.value
    onChange({
      name,
      value: valueEvent,
    })
  }

  return (
    <Container testId={testId} padding={0} className={className}>
      {label && <Label text={`${label}:`} htmlFor={htmlFor} />}
      {Element && (
        <Element
          ariaLabel={ariaLabel}
          name={name}
          onChange={handleChange}
          defaultValue={value}
          className={classNameInput}
        />
      )}
    </Container>
  )
}

export default FormItem
