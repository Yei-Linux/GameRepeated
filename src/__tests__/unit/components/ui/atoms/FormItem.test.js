import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import FormItem from '../../../../../components/ui/atoms/FormItem'

describe('The Form Item component', () => {
  const defaultProps = {
    testId: 'formitem-test',
    htmlFor: 'labelFormItem',
    ariaLabel: 'inputFormItem',
    type: 'input',
    className: 'formItemClass',
    classNameInput: 'inputClass',
    label: '',
    name: 'inputFormItem',
    onChange: jest.fn(),
    value: 'b',
  }

  const setupComponent = (props = defaultProps) =>
    render(<FormItem {...props} />)

  it('Should validate input element', () => {
    const mockOnChange = jest.fn((value) => value)
    const { name, value } = defaultProps
    setupComponent({ ...defaultProps, value: '', onChange: mockOnChange })

    const inputElement = screen.getByRole('textbox', { name: /inputFormItem/i })
    userEvent.type(inputElement, 'b')

    expect(mockOnChange).toBeCalledTimes(1)
    expect(mockOnChange.mock.results[0].value).toStrictEqual({
      name,
      value,
    })
  })

  it('Should validate input type exists', () => {
    setupComponent({
      ...defaultProps,
      type: 'checkbox',
    })

    const inputElement = screen.queryByRole('textbox', {
      name: /inputFormItem/i,
    })

    expect(inputElement).not.toBeInTheDocument()
  })

  describe('FormItem with label', () => {
    it('Should render FormItem with label component', () => {
      const { asFragment } = setupComponent({
        ...defaultProps,
        label: 'LabelUI',
      })

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('FormItem without label', () => {
    it('Should render FormItem without label component', () => {
      const { asFragment } = setupComponent()

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
