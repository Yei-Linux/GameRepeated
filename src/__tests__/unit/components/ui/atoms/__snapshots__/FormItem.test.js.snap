// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Form Item component FormItem with label Should render FormItem with label component 1`] = `
<DocumentFragment>
  <div
    class="formItemClass container--pd0"
    data-testid="formitem-test"
  >
    <label
      class="label"
      for="labelFormItem"
    >
      LabelUI:
    </label>
    <input
      aria-label="inputFormItem"
      autocomplete="off"
      class="inputClass input input--primary input--border1 input--pd1"
      name="inputFormItem"
      type="text"
      value="b"
    />
  </div>
</DocumentFragment>
`;

exports[`The Form Item component FormItem without label Should render FormItem without label component 1`] = `
<DocumentFragment>
  <div
    class="formItemClass container--pd0"
    data-testid="formitem-test"
  >
    <input
      aria-label="inputFormItem"
      autocomplete="off"
      class="inputClass input input--primary input--border1 input--pd1"
      name="inputFormItem"
      type="text"
      value="b"
    />
  </div>
</DocumentFragment>
`;
